### DISCUSSION TITLE
Emacs mode for Rhombus

### DISCUSSION THREAD

**Discussion by kiranandcode** [racket/rhombus#294] [Show and tell]
Not sure if there's already an effort in this direction, but I've been playing around with Rhombus recently, and have gotten tired of having to rely on Python's highlighting as a workaround for making rhombus code pretty.

I've written a basic Emacs mode for Rhombus (very incomplete, but good enough to get the job done for now): https://github.com/Gopiandcode/rhombus-mode

Here are some screenshots on my local machine:

![image-1](https://user-images.githubusercontent.com/23038502/234320127-7bc67188-fd6f-4e47-b77a-2e016ce6b2ae.png)

![image-2](https://user-images.githubusercontent.com/23038502/234320107-ebfc3efd-0ab0-4c0a-8b1b-b70e55e1c433.png)

**Comment by jackfirth** [comment]
You might be interested in some of the work on Racket Mode to make it derive more information from the `#lang` infrastructure. As I understand it, that would in theory make it work and highlight non-racket `#lang`s correctly.

**Reply by kiranandcode** [reply]
Oh, thanks for the reference! Yes, that looks like exactly what I need.

Looking at the racket-mode repository, I'm guessing the functionality that you're referring to is under the branch `lang-lexer`, although it seems like it hasn't been updated in a while and has conflicts with the latest version of racket-mode (https://github.com/greghendershott/racket-mode/compare/master...lang-lexer)

Do you know if it's still being worked on?

**Reply by capfredf** [reply]
I guess the branch is probably `hash-lang`. But why not ask @greghendershott?

**Reply by greghendershott** [reply]
`hash-lang` is the branch. It adds a minor mode, `racket-hash-lang-mode`, which you can enable in a `racket-mode` buffer. The minor mode overrides some font-lock (highlighting) and indent, and sends commands to the Racket Mode back end's Racket process to make use of color-lexer and drraacket:indentation interfaces supplied by a lang.

The branch has been around for quite awhile now. At times I've worked on it pretty intensively. It's non-trivial. Periodically I rebase it onto the main branch (though I haven't done that in awhile).

---

Why haven't I merged it yet? Most Racket Mode features, I dog-food myself pretty regularly. Therefore a bunch of blatant bugs and mis-features get fixed, because they personally bother me. :smile: I am the alpha and beta testing team.

But I don't really dog-food `racket-hash-lang-mode`, because, aside from some rare `#lang scribble/manual`, I am almost always doing things in `racket/base`. At the same time, it's a fairly substantial change, with a high level of I/O going between Emacs and the Racket process used by the Racket Mode back end. So worst case, I merge it, no one uses it for months until I've forgotten details, then suddenly one or more people are heavy users of it and have a ton of problems, and I'm doing a fire-drill when I don't have expect or have time to deal with it. That's my worry.

Instead I guess I'm waiting for a "sponsor". Not financially. I mean someone who really wants and needs this, who will accept it being a little janky at first, who can submit bug reports, and in the future maybe even help me respond to questions from other people trying to use it.

**Reply by samth** [reply]
I've been writing more rhombus recently, and I definitely want `racket-hash-lang-mode`. So I'm happy to dogfood it.

**Reply by greghendershott** [reply]
Great! I just need to figure out a practical concurrency strategy for me to juggle working on both this and `pdb`.

Until you said this: I assumed I'd wrap up `pdb` fully(ish) before really looking at `racket-hash-lang-mode` again.

Instead: I guess I could get `pdb` to some clean(ish) pause point soon, and while waiting to sync up with Robby, reload my brain with `racket-hash-lang-mode`?

**Reply by rfindler** [reply]
One option (dare I suggest...) would be to use racket-hash-lang-mode even for `#lang racket` and then you'd be in the happy dogfooding case again ... :)

**Reply by greghendershott** [reply]
Well.

IME eating my own dog food works well when, sure I need to detour from my intended task to fix a bug in my tool, but at least I'm still eating from the same bowl (same git branch). Shorter, simpler detour. Juggling even just two branches (say main and `pdb`, or main and `hash-lang`) can entail possibly needing to restart Emacs, and using possibly buggy variants of Racket Mode to edit Racket Mode's own Racket source code. And periodically syncing the branches. At three branches... it can start to feel like stepping in dog $NOUN I forgot to pick up earlier _\*cough\*_.

Also: So far, I just don't like the hash-lang approach as much for editing s-expression langs in Emacs. The syntax highlighting is more vanilla, indentation isn't as customizable, other Emacs modes like paredit know nothing about hash-langs, and you need the whole back end running just to edit a plain text file. It's nice that various lexer edge cases are handled correctly compared to the "pile o' regexps etc."... but the overall net benefit for a sexp lang? idk yet.

Of course, some of those deficiencies could be improved through more dog-food detours, and someday probably will, but I do have some other work I'd like to get one, too. :smile:

So TL;DR if someone like Sam is willing/able to alpha test, especially on something like `#lang rhombus`, I think that's the ROI sweet spot in the near future.

To facilitate that: Maybe I should slap some "EXPERIMENTAL!" caveats in the docs and just go ahead and merge the `hash-lang` branch to the main branch. If people don't enable `racket-hash-lang-mode`, it shouldn't break anything (in theory).

**Reply by greghendershott** [reply]
p.s. @samth Meanwhile I did go ahead and rebase `hash-lang` as well as push a few more commits. And somewhat reload my brain. So if you wanted to try that with `#lang rhombus`, I think I'd be ready to field questions or problem reports via Slack or the racket-mode repo issues.

**Comment by Lazerbeak12345** [comment]
While I think this is very cool, I think this shows a greater need for proper LSP in the racket ecosystem. With LSP you'd do this exactly once, then use it across all compatible IDEs, such as VScode, (n)vim (with appropriate plugin), or even emacs (with appropriate plugin).

Further, that "doing once" could be done _in the repo where the language is defined_ - we already have some of this, but it's very DrRacket specific, and doesn't seem general enough to me.

Great work! Looks very readable.

**Reply by greghendershott** [reply]
> LSP

https://github.com/jeapostrophe/racket-langserver wraps functionality provided by things like [`drracket/check-syntax`](https://github.com/racket/drracket/tree/master/drracket-tool-text-lib/drracket) (the library, as opposed to Dr Racket the app) and [`syntax-color`](https://github.com/racket/syntax-color).

It looks like `rhombus-mode` uses this via Emacs `lsp-mode`.

**Reply by kiranandcode** [reply]
When I tried it I wasn't able to get syntax highlighting through `lsp-mode`, so in the end, I wrote my own syntax highlighting rules directly in Emacs --- the repository includes a `lsp-rhombus` file to allow running `lsp-mode` on rhombus buffers, because lsp-mode does show warnings etc. correctly.

**Reply by Lazerbeak12345** [reply]
> > LSP
> 
> https://github.com/jeapostrophe/racket-langserver wraps functionality provided by things like [`drracket/check-syntax`](https://github.com/racket/drracket/tree/master/drracket-tool-text-lib/drracket) (the library, as opposed to Dr Racket the app) and [`syntax-color`](https://github.com/racket/syntax-color).
> 
> It looks like `rhombus-mode` uses this via Emacs `lsp-mode`.

sorry, been busy.

I actually have been using that LSP for a very long time (years? has time gone by that fast?).  My issue with it is it doesn't work on headless servers - while everything else in my setup does (neovim with lspconfig). Cario is still needed at runtime - even though nothing is shown on screen. (https://github.com/jeapostrophe/racket-langserver/issues/45)

>  the racket server relies (heavily) on the editor class text% provided by racket/gui/base, and even just requiring this package causes it to try and initialize a display, even if one won't actually be used. I'm doubtful that there's anything we can do here without basically re-implementing all of the navigation and paragraph features from text%.

This is why I think the problem lies more that racket needs _first party support_ for LSP.

More here: https://github.com/racket/drracket/issues/512#issuecomment-925442956

**Reply by samth** [reply]
I don't think there are any fundamental issues here; racket-mode is built on the drracket infrastructure as well but doesn't need a GUI. I believe the Racket LSP was written that way because it was the easiest approach and it hasn't been changed.

**Reply by greghendershott** [reply]
Partly to support expeditor (an expression editor for the REPL), Matthew added [`syntax-color/color-textoid`](https://github.com/racket/syntax-color/blob/master/syntax-color-lib/syntax-color/color-textoid.rkt) which defines a `color-textoid<%>` interface with a subset of the methods from `text%` that are needed for these purposes -- and which can be implemented without `racket/gui`.

That defines an interface. Each tool -- like expeditor, racket-lang-server, Racket Mode back end -- needs to supply a suitable implementation.

So I think that is probably the direction racket-lang-server would need to take. (Meanwhile an option to use `xvfb-run` would help; maybe that already exists.)

**Reply by rfindler** [reply]
(I'd replied by email with a much less helpful message than @greghendershott 's message above; it didn't actually go into the right place so I just deleted it now. See Greg's!)

**Comment by greghendershott** [comment]
Oh, sorry; I overlooked that. I only glanced in `lsp-rhombus.el`. And even there, if I'd looked more carefully, I would have noticed mentions of both https://github.com/jeapostrophe/racket-langserver and https://github.com/theia-ide/racket-language-server.

Glancing at the latter, the good news is it provides a custom (not standard LSP) notification for fancier coloring from check-syntax information, as well as a custom method for indent. The bad news is the project is in Archived status.

It sounds like you wanted to use both LSP servers, if possible, but that didn't work out?

I guess ideally the extra things from the archived LSP server project could be added to active one, unless there are practical or licensing obstacles.

[Disclaimer: I don't know LSP in any detail, or its use in Emacs (Racket Mode predates LSP).]

**Reply by kiranandcode** [reply]
@greghendershott , ah, no, `lsp-rhombus.el` is just `lsp-racket.el` from lsp-mode, with the substitution 's/racket/rhombus/' (see https://github.com/emacs-lsp/lsp-mode/blob/master/clients/lsp-racket.el)

I included both because `lsp-racket.el` includes both, and I didn't know which one I needed.

(Looking at my Emacs, I guess I was using racket-langserver (the non-archived one))
